### PR TITLE
Add book class

### PR DESCRIPTION
**Pull request by nurlan312**

### CODE REVIEW

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/books")
+@Tag(name = "Book", description = "The Book API")
+public class BookController {
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/BookController.java:18]: @CrossOrigin?

File: src/main/java/kg/peaksoft/ebookm1/config/SpringSecurity.java [hunk 2]
```diff
@@ -40,6 +40,7 @@ protected void configure(HttpSecurity http) throws Exception {
                 .antMatchers("/api/public/**").permitAll()
                 .antMatchers("/v3/api-docs/**", "/swagger-ui/**", "/swagger-ui.html").permitAll()
                 .antMatchers("/api/users/**").permitAll()
+                .antMatchers("/api/books/**").permitAll()
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/config/SpringSecurity.java:43]: Why permit all?

File: src/main/java/kg/peaksoft/ebookm1/dto/book/BookRequest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package kg.peaksoft.ebookm1.dto.book;
+
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import kg.peaksoft.ebookm1.dto.enums.BookLanguage;
+import lombok.Data;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/dto/book/BookRequest.java:8]: Use @Getter and @Setter instead @Data

File: src/main/java/kg/peaksoft/ebookm1/dto/book/BookResponse.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import kg.peaksoft.ebookm1.dto.enums.BookLanguage;
+import kg.peaksoft.ebookm1.entity.book.AudioBook;
+import kg.peaksoft.ebookm1.entity.book.EBook;
+import kg.peaksoft.ebookm1.entity.book.Genre;
+import kg.peaksoft.ebookm1.entity.book.PaperBook;
+import kg.peaksoft.ebookm1.entity.others.Promo;
+import lombok.Data;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/dto/book/BookResponse.java:13]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/book/AudioBook.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package kg.peaksoft.ebookm1.entity.book;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/book/AudioBook.java:10]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/book/Book.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import kg.peaksoft.ebookm1.entity.others.Basket;
+import kg.peaksoft.ebookm1.entity.others.Favorite;
+import kg.peaksoft.ebookm1.entity.others.Promo;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/book/Book.java:14]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/book/EBook.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package kg.peaksoft.ebookm1.entity.book;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/book/EBook.java:10]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/book/Genre.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package kg.peaksoft.ebookm1.entity.book;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/book/Genre.java:9]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/book/PaperBook.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package kg.peaksoft.ebookm1.entity.book;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/book/PaperBook.java:10]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/others/Basket.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package kg.peaksoft.ebookm1.entity.others;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/others/Basket.java:9]: Same

File: src/main/java/kg/peaksoft/ebookm1/repository/BookRepository.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package kg.peaksoft.ebookm1.repository;
+
+import kg.peaksoft.ebookm1.entity.book.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/repository/BookRepository.java:7]: @Service?

File: src/main/java/kg/peaksoft/ebookm1/service/BookService.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        return viewMapper.viewBook(book);
+    }
+
+    public BookResponse updateBook(Long id, BookRequest request) {
+        Book book = repository.findById(id).get();
+        editMapper.updateBook(book, request);
+        return viewMapper.viewBook(repository.save(book));
+    }
+
+    public BookResponse getByIdBook(Long id) {
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/service/BookService.java:34]: Maybe getBookById?

File: src/main/java/kg/peaksoft/ebookm1/service/BookService.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        Book book = repository.findById(id).get();
+        return viewMapper.viewBook(book);
+    }
+
+    public BookResponse deleteByIdBook(Long id) {
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/service/BookService.java:39]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/others/Promo.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package kg.peaksoft.ebookm1.entity.others;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/others/Promo.java:9]: Same

File: src/main/java/kg/peaksoft/ebookm1/entity/others/Favorite.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
+package kg.peaksoft.ebookm1.entity.others;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/others/Favorite.java:9]: Same

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    @Operation(summary = "User with role ADMIN can update")
+    @PreAuthorize("hasAnyAuthority('ADMIN')")
+    @PutMapping("{id}")
+    public BookResponse updateBook(@PathVariable Long id, @RequestBody BookRequest request) {
+        return bookService.updateBook(id, request);
+    }
+
+    @Operation(summary = "Allows all users to get a user by ID")
+    @GetMapping("{id}")
+    public BookResponse getByIdBook(@PathVariable Long id) {
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/BookController.java:38]: getByIdBook?

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+        return bookService.getByIdBook(id);
+    }
+
+    @PreAuthorize("hasAnyAuthority('ADMIN')")
+    @Operation(summary = "User with role ADMIN can delete")
+    @DeleteMapping("{id}")
+    public BookResponse deleteByIdBook(@PathVariable Long id) {
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/BookController.java:45]: deleteByIdBook?

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        return bookService.deleteByIdBook(id);
+    }
+
+    @Operation(summary = "Allows to get all users from the database")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/BookController.java:49]: Allows to get all users?

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Operation(summary = "User with role ADMIN can update")
+    @PreAuthorize("hasAnyAuthority('ADMIN')")
+    @PutMapping("{id}")
+    public BookResponse updateBook(@PathVariable Long id, @RequestBody BookRequest request) {
+        return bookService.updateBook(id, request);
+    }
+
+    @Operation(summary = "Allows all users to get a user by ID")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/BookController.java:36]: users?

**Review by ChyngyzIshen — COMMENTED** [review]
Please pay attention to the comments

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java (hunk 1 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/api/BookController.java:19]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/config/SpringSecurity.java (hunk 2 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/config/SpringSecurity.java]: Corredted

File: src/main/java/kg/peaksoft/ebookm1/dto/book/BookRequest.java (hunk 3 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/dto/book/BookRequest.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/dto/book/BookResponse.java (hunk 4 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/dto/book/BookResponse.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/entity/book/AudioBook.java (hunk 5 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/book/AudioBook.java]: Corredted

File: src/main/java/kg/peaksoft/ebookm1/entity/book/Book.java (hunk 6 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/book/Book.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/entity/book/EBook.java (hunk 7 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/book/EBook.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/entity/book/Genre.java (hunk 8 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/book/Genre.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/entity/book/PaperBook.java (hunk 9 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/book/PaperBook.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/entity/others/Basket.java (hunk 10 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/others/Basket.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/repository/BookRepository.java (hunk 11 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/repository/BookRepository.java]: Corrected to @Repository

File: src/main/java/kg/peaksoft/ebookm1/service/BookService.java (hunk 12 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/service/BookService.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/service/BookService.java (hunk 13 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/service/BookService.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/entity/others/Promo.java (hunk 14 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/others/Promo.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/entity/others/Favorite.java (hunk 15 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/entity/others/Favorite.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java (hunk 16 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/api/BookController.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java (hunk 17 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/api/BookController.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java (hunk 18 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/api/BookController.java]: Corrected

File: src/main/java/kg/peaksoft/ebookm1/api/BookController.java (hunk 19 above)
Comment (nurlan312) [reply src/main/java/kg/peaksoft/ebookm1/api/BookController.java]: Corrected

**Pull request merged**